ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -14,215 +14,308 @@
+    _In_ PPNP_NOTIFY_ENTRY Entry)
+{
+    PAGED_CODE();
+    ObReferenceObject(Entry->DriverObject);
+    Entry->RefCount++;
+}
+
+/**
+ * @brief Calls PnP notification routine and makes some checks to detect faulty drivers
+ */
+static
+CODE_SEG("PAGE")
+VOID
+PiCallNotifyProc(
+    _In_ PDRIVER_NOTIFICATION_CALLBACK_ROUTINE Proc,
+    _In_ PVOID NotificationStructure,
+    _In_ PVOID Context)
+{
+    PAGED_CODE();
+#if DBG
+    KIRQL oldIrql = KeGetCurrentIrql();
+    ULONG oldApcDisable = KeGetCurrentThread()->CombinedApcDisable;
+#endif
+
+    Proc(NotificationStructure, Context);
+
+    ASSERT(oldIrql == KeGetCurrentIrql() &&
+           oldApcDisable == KeGetCurrentThread()->CombinedApcDisable);
+}
+
+static
+CODE_SEG("PAGE")
+_Requires_lock_held_(Lock)
+VOID
+PiProcessSingleNotification(
+    _In_ PPNP_NOTIFY_ENTRY Entry,
+    _In_ PVOID NotificationStructure,
+    _In_ PKGUARDED_MUTEX Lock)
+{
+    PAGED_CODE();
+
+    // the notification may be unregistered inside the procedure
+    // thus reference the driver so we may proceed
+    PiReferencePnpNotifyEntry(Entry);
+
+    // release the lock because the notification routine has to be called without any
+    // limitations regarding APCs
+    KeReleaseGuardedMutex(Lock);
+    PiCallNotifyProc(Entry->PnpNotificationProc, NotificationStructure, Entry->Context);
+    KeAcquireGuardedMutex(Lock);
+
+    PiDereferencePnpNotifyEntry(Entry);
+}
+
+/**
+ * @brief      Delivers the event to all drivers subscribed to
+ *             EventCategoryDeviceInterfaceChange
+ *
+ * @param[in]  Event               The PnP event GUID
+ * @param[in]  InterfaceClassGuid  The GUID of an interface class
+ * @param[in]  SymbolicLinkName    Pointer to a string identifying the device interface name
+ */
+CODE_SEG("PAGE")
+VOID
+PiNotifyDeviceInterfaceChange(
+    _In_ LPCGUID Event,
+    _In_ LPCGUID InterfaceClassGuid,
+    _In_ PUNICODE_STRING SymbolicLinkName)
+{
+    PAGED_CODE();
+
+    PDEVICE_INTERFACE_CHANGE_NOTIFICATION notifyStruct;
+    notifyStruct = ExAllocatePoolWithTag(PagedPool, sizeof(*notifyStruct), TAG_PNP_NOTIFY);
+    if (!notifyStruct)
+    {
+        return;
+    }
+
+    *notifyStruct = (DEVICE_INTERFACE_CHANGE_NOTIFICATION) {
+        .Version = 1,
+        .Size = sizeof(DEVICE_INTERFACE_CHANGE_NOTIFICATION),
+        .Event = *Event,
+        .InterfaceClassGuid = *InterfaceClassGuid,
+        .SymbolicLinkName = SymbolicLinkName
+    };
+
+    DPRINT("Delivering a DeviceInterfaceChange PnP event\n");
+
+    KeAcquireGuardedMutex(&PiNotifyDeviceInterfaceLock);
+
+    PLIST_ENTRY entry = PiNotifyDeviceInterfaceListHead.Flink;
+    while (entry != &PiNotifyDeviceInterfaceListHead)
     {
-        ChangeEntry = CONTAINING_RECORD(ListEntry, PNP_NOTIFY_ENTRY, PnpNotifyList);
-        CallCurrentEntry = FALSE;
+        PPNP_NOTIFY_ENTRY nEntry = CONTAINING_RECORD(entry, PNP_NOTIFY_ENTRY, PnpNotifyList);
+        entry = entry->Flink; // the list entry may be freed
```

Comment:
So, now this just became unsafe. We're not holding a reference to the "next" entry, but `PiProcessSingleNotification` still releases the lock so it could become invalid.
We could:
* take a reference to the next entry here (but then would need extra processing to handle the case where that ends up with a zero refcount)
* have `PiProcessSingleNotification` save the next entry before it derefs the current one and return that
* take the reference to the current entry here rather than in `PiProcessSingleNotification`, so that we can grab the next item before the deref

I'm leaning towards the second option.

Reply:
Have I understood you right, it may become invalid if the next element is freed while the lock is held?
And in such case we want to keep the "current" entry to get the updated `Flink`